### PR TITLE
DEV2-2244 onprem

### PR DESCRIPTION
**Pull request by avnerbarr**

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -5,4 +5,6 @@ dist
 out
 # don't lint nyc coverage output
 coverage
-.eslintrc.js
\ No newline at end of file
+.eslintrc.js
+src/onPremJ.js
```
Comment (dimacodota) [inline .eslintignore:9]: why?

File: .github/workflows/packageOnPrem.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    name: Make vsix
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '16.19'
+      #          runs the binaries downloader script which will be packaged into this plugin
+      - name: pull binaries
+        run: VERSION=${{ inputs.binaryVersion }} ./binaries/downloadBinaries.sh
```
Comment (dimacodota) [inline .github/workflows/packageOnPrem.yml:26]: sync with @assapir - the on prem binary is not released to ga, i think you should download it from the acpl release attachments

File: package.json [hunk 3]
```diff
@@ -96,7 +96,10 @@
   "main": "./out/extension",
   "scripts": {
     "analyze:bundle": "webpack --env analyzeBundle",
-    "vscode:prepublish": "yarn --frozen-lockfile && yarn clear-out && webpack --mode production",
+    "vscode:prepublish": "yarn --frozen-lockfile && yarn clear-out && yarn webpack",
+    "webpack": "yarn webpackGA",
+    "webpackOnPrem": "webpack --mode production --env onPrem",
```
Comment (dimacodota) [inline package.json:101]: `webpackOnPrem`-> `webpack:OnPrem`

File: package.json [hunk 4]
```diff
@@ -96,7 +96,10 @@
...
+    "webpackGA": "webpack --mode production",
```
Comment (dimacodota) [inline package.json:102]: `webpackGA` -> `webpack:GA`

File: package.json [hunk 5]
```diff
@@ -107,6 +110,16 @@
     "test:prepare": "yarn clear-out && tsc && yarn test:copyassets",
     "test": "yarn test:prepare && node ./out/test/runTest.js",
     "vsce:package": "vsce package",
+    "onprem:prep": "yarn onprem:prep:package && yarn onprem:prep:readme",
```
Comment (dimacodota) [inline package.json:113]: would suggest to do all of it in a `buildOnPrem.js` file because it is not readable, this file shold be not coompiled

File: webpack.config.js [hunk 6]
```diff
@@ -63,6 +63,18 @@ const config = {
 };
 
 module.exports = (env) => {
+
+  if (env.onPrem) {
+    console.log("building for onprem");
+    config.entry.extension = "./src/extensionOnPrem.ts";
+    config.plugins.push(new webpack.DefinePlugin({
+      __ONPREM__: JSON.stringify(true)
```
Comment (dimacodota) [inline webpack.config.js:71]: why here and not in the action ?

File: src/binary/paths.ts [hunk 7]
```diff
@@ -5,68 +5,103 @@ import {
...
 const BUNDLE_SUFFIX = getBundleSuffix();
 
 export async function setBinaryRootPath(
   extensionContext: vscode.ExtensionContext
 ): Promise<void> {
-  binaryRootPath =
+  if (ONPREM) {
+    const base = `binaries/${getArch()}-${getPlatform()}`;
+    BINARY_ROOT_PATH = path.join(extensionContext.extensionPath, base);
+    await makeExecutable(BINARY_ROOT_PATH);
```
Comment (dimacodota) [inline src/binary/paths.ts:21]: already implemented use  this https://github.com/codota/tabnine-vscode/blob/dc35e8a59c7ecd0e074c893fc0a58f5bfb3e1c20/src/binary/binaryFetcher/bundleDownloader.ts#L88

File: src/binary/paths.ts [hunk 8]
```diff
@@ -5,68 +5,103 @@ import {
...
+  }
+  const files = await fs.readdir(bundleDirectory);
+  console.log("making files executable", files);
+  await Promise.all(
+    files.map((file) => fs.chmod(path.join(bundleDirectory, file), 0o755))
+  );
+}
+
+export function bundledTabnineBinaryPath(): string {
+  return path.join(<string>BINARY_ROOT_PATH, binaryName());
```
Comment (dimacodota) [inline src/binary/paths.ts:49]: BINARY_ROOT_PATH can be undefined

File: src/binary/paths.ts [hunk 9]
```diff
@@ -5,68 +5,103 @@ import {
...
 const ARCHITECTURE = getArch();
 const SUFFIX = getSuffix();
 const BUNDLE_SUFFIX = getBundleSuffix();
 
 export async function setBinaryRootPath(
   extensionContext: vscode.ExtensionContext
 ): Promise<void> {
-  binaryRootPath =
+  if (ONPREM) {
+    const base = `binaries/${getArch()}-${getPlatform()}`;
```
Comment (dimacodota) [inline src/binary/paths.ts:19]: why ? all this is already implemented, you can use this https://github.com/codota/tabnine-vscode/blob/dc35e8a59c7ecd0e074c893fc0a58f5bfb3e1c20/src/binary/paths.ts#L29

File: src/binary/runBinary.ts [hunk 10]
```diff
@@ -13,14 +14,19 @@ export default async function runBinary(
 
   const context = getTabnineExtensionContext();
   const args: string[] = [
-    "--client=vscode",
+    ONPREM ? "--client=vscode_on_prem" : "--client=vscode",
     "--no-lsp=true",
     tabnineExtensionProperties.logFilePath
       ? `--log-file-path=${tabnineExtensionProperties.logFilePath}`
       : null,
+    ONPREM ? "--onprem=true" : null,
```
Comment (dimacodota) [inline src/binary/runBinary.ts:22]: i think it is irelevant

File: src/binary/runBinary.ts [hunk 11]
```diff
@@ -13,14 +14,19 @@ export default async function runBinary(
...
+    ONPREM ? "--no_bootstrap" : null,
```
Comment (dimacodota) [inline src/binary/runBinary.ts:23]: i think it is irelevant

File: src/binary/runProcess.ts [hunk 12]
```diff
@@ -22,6 +22,12 @@ export function runProcess(
     return require("./mockedRunProcess").default() as BinaryProcessRun;
   }
   report(EventName.START_BINARY);
+  console.log(
```
Comment (dimacodota) [inline src/binary/runProcess.ts:25]: remove

File: src/extensionOnPrem.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+function initStartup(context: vscode.ExtensionContext): void {
+  setTabnineExtensionContext(context);
+}
+
+async function backgroundInit(context: vscode.ExtensionContext) {
+  await setBinaryRootPath(context);
+  await initBinary();
+  notifyBinaryAboutWorkspaceChange();
```
Comment (dimacodota) [inline src/extensionOnPrem.ts:35]: this is irelevant for ton prem, it is being used in the right side panel

File: src/notifications/pollNotifications.ts [hunk 14]
```diff
@@ -32,6 +33,9 @@ export function cancelNotificationsPolling(): void {
 export async function doPollNotifications(
   context: vscode.ExtensionContext
 ): Promise<void> {
+  if (ONPREM) {
```
Comment (dimacodota) [inline src/notifications/pollNotifications.ts:36]: irelevant you do not call pollNotifications

File: src/commandsHandler.ts [hunk 15]
```diff
@@ -13,13 +14,14 @@ export const SAVE_SNIPPET_COMMAND = "Tabnine.saveSnippet";
...
   context: ExtensionContext
 ): Promise<void> {
-  context.subscriptions.push(
-    commands.registerCommand(CONFIG_COMMAND, openHub(StateType.PALLETTE))
-  );
-
-  context.subscriptions.push(
-    commands.registerCommand(STATUS_BAR_COMMAND, handleStatusBar(context))
-  );
+  if (!ONPREM) {
```
Comment (dimacodota) [inline src/commandsHandler.ts:17]: remove it from here and just do not call `registerCommands` from the extensionOnPrem.ts

File: src/reports/reporter.ts [hunk 16]
```diff
@@ -26,6 +27,9 @@ export function initReporter(
   if (inTestMode) {
     return;
   }
+  if (ONPREM) {
```
Comment (dimacodota) [inline src/reports/reporter.ts:30]: change the `reporter` to be `TelemetryReporter | undefined` and then check if it is not `undefined` instead of duplicate the check

File: src/statusBar/StatusBarData.ts [hunk 17]
```diff
@@ -56,6 +58,16 @@ export default class StatusBarData {
   }
 
   private updateStatusBar() {
+    if (ONPREM) {
+      const issueText = this._text ? `: ${this._text}` : "";
+      const limited = this._limited ? ` ${LIMITATION_SYMBOL}` : "";
```
Comment (dimacodota) [inline src/statusBar/StatusBarData.ts:63]: _limited is irelevant

File: src/extensionOnPrem.ts [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+import installAutocomplete from "./autocompleteInstaller";
+import { handleSelection, notifyBinaryAboutWorkspaceChange } from "./extension";
+
+export async function activate(
+  context: vscode.ExtensionContext
+): Promise<void> {
+  console.log("*****THIS IS AN ONPREM BUILD****");
+  void initStartup(context);
+  handleSelection(context);
+  registerStatusBar(context);
```
Comment (dimacodota) [inline src/extensionOnPrem.ts:21]: i would override this function and and create an onprem `registerStatusBar` function instead of modifying the `updateStatus` function

File: webpack.config.js [hunk 19]
```diff
@@ -63,6 +63,18 @@ const config = {
 };
 
 module.exports = (env) => {
+
+  if (env.onPrem) {
+    console.log("building for onprem");
+    config.entry.extension = "./src/extensionOnPrem.ts";
```
Comment (dimacodota) [inline webpack.config.js:69]: this is an uncommon practice to do it , you can just modify the https://github.com/codota/tabnine-vscode/blob/dc35e8a59c7ecd0e074c893fc0a58f5bfb3e1c20/webpack.config.js#L14
and do something like ``"./src/extension${env.onPrem ? "OnPrem" : ""}.ts"``

File: src/reports/reporter.ts (hunk 16 above)
Comment (avnerbarr) [reply src/reports/reporter.ts:30]: then you would be led to believe that it should be defined. this check makes the intention obvious and explicit

File: src/commandsHandler.ts (hunk 15 above)
Comment (avnerbarr) [reply src/commandsHandler.ts:17]: some of the commands need to be registered

File: src/notifications/pollNotifications.ts (hunk 14 above)
Comment (avnerbarr) [reply src/notifications/pollNotifications.ts:36]: it is called on every selection suggestion

File: src/extensionOnPrem.ts (hunk 13 above)
Comment (avnerbarr) [reply src/extensionOnPrem.ts:35]: I'll remove. it isn't aware what the point of this class is

File: src/binary/runBinary.ts (hunk 11 above)
Comment (avnerbarr) [reply src/binary/runBinary.ts:23]: we are still passing the flags. if it gets changed I will change here

File: src/binary/runBinary.ts (hunk 10 above)
Comment (avnerbarr) [reply src/binary/runBinary.ts:22]: not yet

File: src/binary/paths.ts (hunk 9 above)
Comment (avnerbarr) [reply src/binary/paths.ts:19]: there is no version . it is bundled flat

File: src/binary/paths.ts (hunk 8 above)
Comment (avnerbarr) [reply src/binary/paths.ts:49]: this is the current convention. if it could happen now, it could happen before. the design isn't great to have dangling references

**Comment by dimacodota** [comment]
General notes: 
 - sync with @assapir  on the onprem binary build location
 - since you already created a different entry point for the onprem extension, idd try to manage the "onprem" decisions on this level and not  on the functional level where you  need to add the `if(onprem)` everywhere

File: webpack.config.js (hunk 6 above)
Comment (avnerbarr) [reply webpack.config.js:71]: GitHub action? How could it be in the action, when the web pack is doing the replacements

File: package.json (hunk 5 above)
Comment (avnerbarr) [reply package.json:113]: because there are shell commands here and this mechanism allows for running from outside individually

File: .eslintignore (hunk 1 above)
Comment (avnerbarr) [reply .eslintignore:9]: irrelevant now. will remove it

File: src/binary/runProcess.ts (hunk 12 above)
Comment (avnerbarr) [reply src/binary/runProcess.ts:25]: what's the problem with this log

File: package.json [hunk 20]
```diff
@@ -1,7 +1,7 @@
 {
   "publisher": "TabNine",
   "name": "tabnine-vscode",
-  "version": "3.6.38",
+  "version": "3.6.40",
```
Comment (dimacodota) [inline package.json:4]: revert it and use the bump action to release

File: src/binary/paths.ts (hunk 9 above)
Comment (dimacodota) [reply src/binary/paths.ts:20]: ok , but all the path logic is already implemented there, you can extract it or something

File: src/binary/paths.ts (hunk 8 above)
Comment (dimacodota) [reply src/binary/paths.ts:38]: i meant the casting you are doing here, it will be more clear to adjust the return type and this casting

File: src/binary/paths.ts [hunk 21]
```diff
@@ -5,68 +5,92 @@ import {
...
   } catch (err) {
     // Exception is thrown if the path already exists, so ignore error.
   }
 }
 
+export function bundledTabnineBinaryPath(): string {
+  return path.join(<string>BINARY_ROOT_PATH, binaryName());
+}
+
+function binaryName(): string {
```
Comment (dimacodota) [inline src/binary/paths.ts:41]: getSuffix

File: src/binary/binaryFetcher/existingVersionHandler.ts [hunk 22]
```diff
@@ -1,12 +1,20 @@
...
 import { getRootPath, versionPath } from "../paths";
+import { setDirectoryFilesAsExecutable } from "../utils";
 
 export default async function handleExistingVersion(): Promise<string | null> {
   try {
     const versionPaths = await fs.readdir(getRootPath());
+    await Promise.all(
+      versionPaths
+        .map((version) => path.dirname(versionPath(version)))
+        .filter((p) => existsSync(p))
```
Comment (dimacodota) [inline src/binary/binaryFetcher/existingVersionHandler.ts:15]: why sync?

File: src/binary/binaryFetcher/existingVersionHandler.ts [hunk 23]
```diff
@@ -1,12 +1,20 @@
...
+        .map(async (p2) => setDirectoryFilesAsExecutable(p2))
```
Comment (dimacodota) [inline src/binary/binaryFetcher/existingVersionHandler.ts:16]: change to `map(setDirectoryFilesAsExecutable)`

File: src/commandsHandler.ts (hunk 15 above)
Comment (dimacodota) [reply src/commandsHandler.ts:17]: what commands? enableSaveSnippetContext is not releavant

File: src/extensionOnPrem.ts [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+async function backgroundInit(context: vscode.ExtensionContext) {
+  await setBinaryRootPath(context);
+  await initBinary();
+
+  setDefaultStatus();
+  void registerCommands(context);
+  await installAutocomplete(context);
+  vscode.languages.registerHoverProvider(
```
Comment (dimacodota) [inline src/extensionOnPrem.ts:39]: this not relevant to onprem

File: src/binary/binaryFetcher/existingVersionHandler.ts (hunk 22 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/existingVersionHandler.ts:15]: use https://github.com/codota/tabnine-vscode/blob/8e1cdc07931034dc78c552327c2b6ccee2e12e53/src/utils/file.utils.ts#L28

File: src/binary/binaryFetcher/existingVersionHandler.ts (hunk 23 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/existingVersionHandler.ts:17]: must say it is a bit strange to manage it here and the downloading in a different part of the code, id suggest to to make the files  executable after  you bundle them, somewhere in the ci. and keep the  downloading/bundling/chmod together because they are related

Comment (avnerbarr) [reply src/binary/binaryFetcher/existingVersionHandler.ts:17]: Not possible. Chmod is platform dependent and that is why the downloader does it as well. The chmod has to occur somewhere and this is the first place it is possible. I have to admit that the "duplicated" code was simpler by 10x and that is why i did it. the amount of changes required in order to not "duplicate" the path.join trivial code are immense in comparison  and not worth the "smell" of the dup. 
In other words IMO the cost of "duplicating" trivial code of 1 line vs changes that need to be propagated to tens of places are definitely not worth it

Comment (dimacodota) [reply src/binary/binaryFetcher/existingVersionHandler.ts:17]: 1 -the downloader does the chmod bc the download happens after the installation.
2 - it is posible that we we did in the "bundled version, you can see it here https://github.com/codota/TabNine/blob/8ba545f20dac8de3b0fc0c6acf377a29e89bbcb4/dl_binaries.sh#LL23C30-L23C30

3 - disagree that this is a trivial code . it is is not bc you need to follow why it is here

**Review by dimacodota — APPROVED** [review]
lgtm, although I don't fully agree with some parts here. I do understand the time frame and I think it will be easier to fix after that

File: src/binary/binaryFetcher/existingVersionHandler.ts (hunk 23 above)
Comment (dimacodota) [reply src/binary/binaryFetcher/existingVersionHandler.ts:17]: btw, the duplication was not in `path.join` it was in the calculation of the arch and the platform , which  just not makes any sense

**Comment by dimacodota** [comment]
you do need to modify the plugin id in the package.json

File: webpack.config.js (hunk 19 above)
Comment (avnerbarr) [reply webpack.config.js:69]: there is no env object there
